dask/distributed/scheduler.py

```diff
@@ -0,0 +1,597 @@
+from __future__ import print_function
+
+import zmq
+import socket
+import dill
+import uuid
+from collections import defaultdict
+import itertools
+from multiprocessing.pool import ThreadPool
+import random
+from datetime import datetime
+from threading import Thread, Lock
+from contextlib import contextmanager
+from toolz import curry, partial
+from ..compatibility import Queue, unicode
+try:
+    import cPickle as pickle
+except ImportError:
+    import pickle
+
+from ..core import get_dependencies, flatten
+from .. import core
+from ..async import finish_task, start_state_from_dask as dag_state_from_dask
+
+with open('log.scheduler', 'w') as f:  # delete file
+    pass
+
+def log(*args):
+    with open('log.scheduler', 'a') as f:
+        print(*args, file=f)
+
+@contextmanager
+def logerrors():
+    try:
+        yield
+    except Exception as e:
+        log('Error!', str(e))
+        raise
+
+class Scheduler(object):
+    """ Disitributed scheduler for dask computations
+
+    Parameters
+    ----------
+
+    hostname: string
+        hostname or IP address of this machine visible to outside world
+    port_to_workers: int
+        Port on which to listen to connections from workers
+    port_to_clients: int
+        Port on which to listen to connections from clients
+    block: bool
+        Whether or not to block the process on creation
+
+    State
+    -----
+
+    workers - dict
+        Maps worker identities to information about that worker
+    who_has - dict
+        Maps data keys to sets of workers that own that data
+    worker_has - dict
+        Maps workers to data that they own
+    data - dict
+        Maps data keys to metadata about the computation that produced it
+    to_workers - zmq.Socket (ROUTER)
+        Socket to communicate to workers
+    to_clients - zmq.Socket (ROUTER)
+        Socket to communicate with users
+    """
+    def __init__(self, port_to_workers=None, port_to_clients=None,
+                 hostname=None, block=False):
+        self.context = zmq.Context()
+        hostname = hostname or socket.gethostname()
+
+        # Bind routers to addresses (and create addresses if necessary)
+        self.to_workers = self.context.socket(zmq.ROUTER)
+        if port_to_workers is None:
+            port_to_workers = self.to_workers.bind_to_random_port('tcp://*')
+        else:
+            self.to_workers.bind('tcp://*:%d' % port_to_workers)
+        self.address_to_workers = ('tcp://%s:%d' % (hostname, port_to_workers)).encode()
+
+        self.to_clients = self.context.socket(zmq.ROUTER)
+        if port_to_clients is None:
+            port_to_clients = self.to_clients.bind_to_random_port('tcp://*')
+        else:
+            self.to_clients.bind('tcp://*:%d' % port_to_clients)
+        self.address_to_clients = ('tcp://%s:%d' % (hostname, port_to_clients)).encode()
```

Comment:
I've gone about this in a few ways. One is to allow protocol, host, and port to be specified separately. I've only done this where these things can be specified via config files. I would imagine it would be quite tedious in a function/constructor signature.

The opposite approach is to make the argument the full 'tcp://ip:port' URL the argument, and put the more fine-grained configuration at a higher level, or in a helper function if you don't want people to type out URLs.

Another option, and probably not actually a good idea, is to make the URL the argument, but allow it to be specified with abbreviations - integers mean `tcp://*:%i' % n`, str without `://` imply tcp, etc.